<a href="https://colab.research.google.com/github/knail1/bayes/blob/main/Isye6420_hw3_q1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## HW3 ISYE 6420 Fall 2021
### student: Omer Ansari (oansari3)
### date - 10/10/21

## Q1

**(a) compare the Bayes estimator for $\theta$ with the MLE.**

*Approach Summary: We first measure the mean of the MLE. Second, we figure out the posterior by multiplying the prior with the likelihood. Third, we observe the emerging distribution of the posterior, and use that distributions mean to figure out the Bayes' estimator (mean). Fourth, we compare this bayes' estimator with the likelihood in terms of percentage difference*

- Prior: $\pi(\theta) = \frac{1}{\sqrt{\theta}}$ for $0<\theta<\infty$ ---- A.1
- likelihood: $f(x|\theta) = Poi(\theta) = \frac{\theta^{x}}{x!}e^{-\theta}$ -- A.2
- observed data X = 1,2,0,1

First, we calculate the mean of the likelihood to calculate the frequentist approach:
- mean of likelihood (frequentist approach) for $Poi(\theta):$
  - $\widehat{\theta}_{MLE} = \widehat{X} $
  - and $\widehat{X} = \frac{1+2+0+1}{4} = 1$
  - thus $\widehat{\theta}_{MLE} = 1$ ---- A.3

Next, we calculate posterior using:

$\pi(\theta|x) = f(x|\theta) * \pi(\theta)$ ---- A.3.1

now we calculate likelihood _after_ the observations:

We have 4 observed values (1,2,0,4), let's calculate likelihood:

- $\pi(\theta|1) = \theta^{1-0.5}e^{-\theta}$ --- A.4
- $\pi(\theta|2) = \frac{1}{2}\theta^{2-0.5}e^{-\theta}$ --- A.5
- $\pi(\theta|0) = \theta^{0-0.5}e^{-\theta}$ --- A.6
- $\pi(\theta|1) = \theta^{1-0.5}e^{-\theta}$ --- A.7

Plugging A.4, A.5, A.6, A.7:

Likelihood = $\pi(\theta|1) . \pi(\theta|2) . \pi(\theta|0) . \pi
(\theta|1)$

$= \frac{1}{2}\theta^{4}e^{-4\theta}$ ---- A.8


Plugging A.1, and and A.8 into A.3.1, the **posterior** is:


$$\pi(\theta|x) = \frac{1}{2}\theta^{3.5}e^{-4\theta}$$ --- A.9

The posterior can be observed to be a Gamma distribution.
As it were,
$\theta = Ga(\alpha, \beta)$, where

$Ga \approx \theta^{\alpha-1}e^{-\beta\theta}$

that equates to:

$\alpha - 1 = 3.5 $ and 
$\beta = 4$

mean of $\theta = Ga(\alpha, \beta)$ is $\frac{\alpha}{\beta}$

thus
$\widehat{\theta}_{bayes} = 4.5/4 = 1.125$


*Answer: The Bayes Estimator is 12.5% higher than the MLE.*


--------



**(b) Compute numerically a 95% equitailed credible set.**

- With the posterior distribution found in (a), you want to find the points L, U such that P(theta < L) = P(theta > U), and those add up to 5%.

-----

In [ ]:
import pymc3
from scipy.stats import gamma
a = 3.5
gamma_data = gamma.rvs(a, scale, size=1000)
L = gamma.ppf(0.05, a, loc=0, scale=1)
R = gamma.ppf(0.95, a, loc=0, scale=1)

print("The credible set for 0.95 equitail is from {} to {}".format(L,R))

The credible set for 0.95 equitail is from 1.0836749546490285 to 7.033570224670084


--------
**(c) Compute numerically a 95% HPD credible set.**




In [ ]:
import pymc3
from scipy.stats import gamma

#first generate data from the distribution
# recall: alpha = 3.5, and beta = 4
a = 3.5
scale = 1/4
#gamma_data = gamma.rvs(a, scale, size=1000)

hpd_limits = []
hpd_limits = pymc3.hdi(gamma_data, hdi_prob=.95)

print("HPD credible set: {}".format(hpd_limits))

HPD credible set: [0.51779319 7.27261705]


------

**(d) Numerically find the mode of the posterior, that is, MAP estimator of $\theta$**

For $Ga(\alpha,\beta)$, mode = $\frac{\alpha-1}{\beta}$
- $\alpha$ = 3.5
- $\beta$ = 4

Mode = 2.5/4 = 0.625



-----
**(e) If you test hypothesis $H_{0}: \theta \ge 1$ vs $H_{1}: \theta < 1 $ based on the posterior, which hypothesis will be favored?**

posterior:
$$\pi(\theta|x) = \frac{1}{2}\theta^{3.5}e^{-4\theta}$$ 


- we can calculate two CDFs of this posterior for :
  - $H_{0}$ : range of $\theta$ from 1 to $\infty$ 
  - $H_{1}$ : range of $\theta$ from -$\infty$ to 1


In [ ]:
gammacdf = gamma.cdf(gamma_data, a, loc=0, scale =1)

H0_cdf = 1 - gammacdf[1]
#print(gammacdf.shape)
print("H0_cdf = {}".format(H0_cdf))

#H1_cdf = gammacdf[0] + gammacdf[1]
H1_cdf = gammacdf[1]

print("H1_cdf = {}".format(H1_cdf))

H0_cdf = 0.6716818684894892
H1_cdf = 0.3283181315105108


We can see that
- $H_{0} : \theta = 0.6716719$
- $H_{1} : \theta = 0.3283181$

therefore $H_{0}$ is superior.





--------

#Q2 

**data:**
- lady guesses 15/16 correctly
  - n = 16 -- A.1
  - x = 15 -- A.2
- p = probability of lady guessing correctly
- probability of $H_{0}$ is p =0.5 (she's just guessing)
- probability of $H_{1}$ is p > 0.5 (she has ESP!)
- prior $\pi(p) = \pi_{0}\sigma_{0.5} + \pi_{1}U(0.5,1)$
  - $\pi_{0}$ = 0.95 --- A.3
  - $\pi_{1}$ = 0.05 --- A.4
- likelihood $f(x|p) = Bin(15,p) = $ $P(X \geq 15 | H_{0}$)
  - $\binom{16}{15}0.5^{15}0.5^{1} + \binom{16}{16}0.5^{16}0.5^{0} = 0.0002594 $

--

*(a) find the posterior probabilities of hypothesis, p0 and p1, and Bayes Factor:*

first some equations:

$p_{0} = P(H_{0}|X) = [ 1 + \frac{\pi_{1}}{\pi_{0}}. \frac{m_{1}(x)}{f(x|0.5)}]^{-1}$ ---- P.1

$B_{01} = \frac{f(x|0.5)}{m_{1}(x)}$. ---- P.2

$f(x|p) = \binom{n}{x}p^{x}(1-p)^{n-x}$ ---- P.3

$m_{1}(x) = \int_{0.5}^{1}f(x|p)2dp$ ----- P.4

First, we integrate P.4:
- plug in A.1 and A.2 in P.3:

$f(x|p) = f(15|p) = \binom{16}{15}p^{16}(1-p)^{16-15}$
$= 16(p^{15}(1-p))$ ---- P.5

- plug P.5 , and A.2 into P.4:

$m_{1}(15) = \int_{0.5}^{1}16(p^{15}(1-p)2dp$ ----- P.4

$ = 32\int_{0.5}^{1}(p^{15}(1-p)dp$ 

$ = 32\left | \frac{p^{16}}{16}-\frac{p^{17}}{17}\right | $ over 0.5 to 1

$ =  32* ( (\frac{1}{16} - \frac{1}{17}) - (\frac{0.5^{16}}{16} - \frac{0.5^{16}}{17}))$









In [ ]:
term1 = 1/16 - 1/17
term2 = (0.5**16)/16 - (0.5**17)/17
m1 = 32*(term1-term2)
print("m1(15) = {} ------ P.5 ".format(m1)) # ---- P.5

m1(15) = 0.11763090245863973 ------ P.5 


now, let's evaluate f(x|0.5), using A.2 and P.5:

for x = 15:

$f(x|0.5) = 16((0.5)^{15}(1-0.5))$

In [ ]:
fx0_5 = 16*(0.5**15)*(1-0.5)
print("f(x|0.5) = {} --------P.6 ".format(fx0_5)) # ---- P.6

f(x|0.5) = 0.000244140625 --------P.6 


recall that P.1 is:
$p_{0} = P(H_{0}|X) = [ 1 + \frac{\pi_{1}}{\pi_{0}}. \frac{m_{1}(x)}{f(x|0.5)}]^{-1}$ ---- P.1

- plug in A.3, A.4, P.5, and P.6  into P.1:

$p_{0} = [1 + \frac{0.05}{0.95}. \frac{0.1176309}{0.0002441}]^{-1}$

as well as evaluating $p_{1} = 1 - p_{0}$ : 

In [ ]:
term3 = 0.05/0.95
term4 = m1/fx0_5
p0 = (1 + term3*term4)**(-1)
print("p0 = {} ------ P.7".format(p0))
p1 = 1 - p0
print("p1 = {} ------ P.8".format(p1))

p0 = 0.037938071677115284 ------ P.7
p1 = 0.9620619283228847 ------ P.8


therefore, to 7 significant digits (we are using the same number of sig digits illustrated for p-value in Q2 hint)

$p_{0} = 0.0379381$. ---- P.7

$p_{1} = 0.9620619$ ---- P.8


--
recall: $B_{01} = \frac{f(x|0.5)}{m_{1}(x)}$. ---- P.2

now, we compute Bayes factor in favor of $H_{0}$ by plugging in P.6 and P.5 into P.2:

$B_{01} = \frac{0.000244140625}{0.11763090245863973}$


In [ ]:
B01 = 0.000244140625/0.11763090245863973
print("B01 = {} ---- P.9".format(B01))

# alternate : B01 = (p0/p1)/(pi0/pi1)
# using P.7, P.8, A.3, A.4:
num = 0.037938071677115284/0.9620619283228847
den = (0.95)/(0.05)
B01_alt = num / den
#print(B01_alt)


B01 = 0.0020754803363499013 ---- P.9
0.0020754803363499013


$B_{01} = 0.0020755$


-------

**part (b) what is the assessment of H1 according to Jeffrey's scale, is the experiment convincing that the lady possesses ESP**

for this we use this equation:

$log_{10}B_{10} = -log_{10}(B_{01})$ --- P.10

plug P.9 into P.10:


In [ ]:
import math
lg_b10 = -math.log10(B01)

print("log(B10) = {}".format(lg_b10))

log(B10) = 2.6828813768868525


according to Jeffrey's scale:

when $log_{10}B_{10}(x) > 2$ the evidence against $H_{0}$ is decisive. This means, the experiment **IS** convincing that the lady possess ESP, even with Bayes treatment.